vmdb/app/models/mixins/naming_index_mixin.rb

```diff
@@ -0,0 +1,20 @@
+module NamingIndexMixin
+  extend ActiveSupport::Concern
+
+  NAMING = 'naming'
+
+  included do
+    has_many :naming_indexes, :as => :resource, :dependent => :destroy, :class_name => "CustomAttribute", :conditions => {:source => NAMING}
+  end
+
+  def next_naming_index(name)
+    with_lock do
```

Comment:
First off, sorry for the length, but... @bzwei, @matthewd

> Then why don't we switch to use with_lock, or at least stick with with_lock here?

Because with_timeout doesn't have the timeout and it doesn't have the simplicity of saying `:shared` or `:exclusive`.  The timeout is what makes our lock code "safer" (safer as in safety net).

Additionally, for me, the parameter on with_lock is confusing because the caller has to give a raw SQL string when they want to do a different kind of lock, and then the caller is responsible for SQL fragments, which is generally avoided.  Also, saying lock(true) is meaningless when reading the code.  What does a "true" lock mean?  For that matter what is a "false" lock? o_O  We felt that by providing :shared or :exclusive as the parameter and then let the lock method figure out the actual SQL it was clearer from the calling side.  This is what the ActiveRecordQueryParts module is for, as well as to support multiple databases like we used to in the past.  My long term goal was to push that part into Rails, but alas, I never got there.

> The database should deal with deadlocks itself, so do we have situations where we do other potentially-slow things while holding a lock? Things that can be safely/usefully aborted?

Opening a transaction, acquiring a lock, and then holding it forever is totally possible with with_lock.  Databases don't determine that to be a deadlock situation; it's a normal mutual exclusion setup, so the database will just stay in that state forever.  A deadlock is process X holds lock A, then asks for lock B, while process Y holds lock B and asks for lock A.  That isn't what's happening here.  Here we have process X holds lock A, then process Y asks for lock A and just waits in line.  So if you do something like `Account.first.with_lock { sleep(1.day) }` or `Account.first.with_lock { execute "SELECT pg_sleep(seconds)" }`, the database will hold that lock for a day, and you're just stuck.

In fairness, on the database configuration side you can set a parameter for the max lock holding time, but it's for the entire database.  I think ours is set for 10 minutes (maybe even more)...this is because an especially long query in reporting is technically valid, and is sometimes necessary (though we usually discourage it).  So, a client side timeout is useful when we know it should never take longer than some value, because it lets us limit a particular query in Ruby and not have it lock up the entire system (at least for the timeout length).

> What's the history of the timeout?

I searched the source history and found the timeout was added in the below commit in zone.rb, and was later extracted out to a mixin to be used on other models.

```
commit cd9eca9469db13092287278966ca2c831a069d8b
Author: Gregg Tanzillo <gtanzillo@manageiq.com>
Date:   Wed Jun 17 00:35:34 2009 +0000

    BugzID: 5033
    -Removed all shared locks fro queue code because they were not necessary.
    -Changed exclusive lock to shared lock in lock_in_my_zone because the exclusive locking was excessive.
    -Added a 60 timeout around block in lock_in_my_zone to prevent blocking other processes when the execution does not return.
```
ASIDE: ping @gtanzillo (hi!)

Unfortunately, the trail ends there and I can't find more info on the issue that was in question.  My guess is something went crazy, the lock was held for an inordinate amount of time and locked up the entire system.  These locks are heavily used on startup when the database needs to seed, where multiple workers are trying to seed the system.  If one locks up permanently, the system effectively never starts.

> lying debug message

Is it?  How so?